ping/main.go

```diff
@@ -0,0 +1,284 @@
+	// the test immediately.
+	initCtx.MustWaitAllInstancesInitialized(ctx)
+
+	// 🐥  Now all instances are ready for action.
+	//
+	// Note: In large test runs, the scheduler might take a few minutes to
+	// schedule all instances in a cluster.
+
+	// In containerised runs (local:docker, cluster:k8s runners), Testground
+	// instances get attached two networks:
+	//
+	//   * a data network
+	//   * a control network
+	//
+	// The data network is where standard test traffic flows. The control
+	// network connects us ONLY with the sync service, InfluxDB, etc. All
+	// traffic shaping rules are applied to the data network. Thanks to this
+	// separation, we can simulate disconnected scenarios by detaching the data
+	// network adapter, or blocking all incoming/outgoing traffic on that
+	// network.
+	//
+	// We need to listen on (and advertise) our data network IP address, so we
+	// obtain it from the NetClient.
+	ip := initCtx.NetClient.MustGetDataNetworkIP()
+
+	var security libp2p.Option
+	switch secureChannel {
+	case "noise":
+		security = libp2p.Security(noise.ID, noise.New)
+	case "secio":
+		security = libp2p.Security(secio.ID, secio.New)
+	case "tls":
+		security = libp2p.Security(tls.ID, tls.New)
+	}
+
+	// ☎️  Let's construct the libp2p node.
+	listenAddr := fmt.Sprintf("/ip4/%s/tcp/0", ip)
+	host, err := libp2p.New(ctx,
+		security,
+		libp2p.ListenAddrStrings(listenAddr),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to instantiate libp2p instance: %w", err)
+	}
+
+	// 🚧  Now we instantiate the ping service.
+	//
+	// This adds a stream handler to our Host so it can process inbound pings,
+	// and the returned PingService instance allows us to perform outbound pings.
+	ping := ping.NewPingService(host)
+
+	// Record our listen addrs.
+	runenv.RecordMessage("my listen addrs: %v", host.Addrs())
+
+	// Obtain our own address info, and use the sync service to publish it to a
+	// 'peersTopic' topic, where others will read from.
+	var (
+		id = host.ID()
+		ai = &peer.AddrInfo{ID: id, Addrs: host.Addrs()}
+
+		// the peers topic where all instances will advertise their AddrInfo.
+		peersTopic = sync.NewTopic("peers", new(peer.AddrInfo))
+
+		// initialize a slice to store the AddrInfos of all other peers in the run.
+		peers = make([]*peer.AddrInfo, 0, runenv.TestInstanceCount-1)
+	)
+
+	// Publish our own.
+	initCtx.SyncClient.MustPublish(ctx, peersTopic, ai)
+
+	// Now subscribe to the peers topic and consume all addresses, storing them
+	// in the peers slice.
+	peersCh := make(chan *peer.AddrInfo)
+	sctx, cancel := context.WithCancel(ctx)
+	sub := initCtx.SyncClient.MustSubscribe(sctx, peersTopic, peersCh)
+
+	// Receive the expected number of AddrInfos.
+	for len(peers) < cap(peers) {
+		select {
+		case ai := <-peersCh:
+			if ai.ID == id {
+				continue // skip over ourselves.
+			}
+			peers = append(peers, ai)
+		case err := <-sub.Done():
+			return err
+		}
+	}
+	cancel()
```

Comment:
nit: why do we cancel at the end here? if this is just about releasing resources we might avoid it as (or comment it explicitly as such), otherwise it might confuse the reader thinking this is needed for something more in the correct operation of the synchronization test.

Reply:
This cancels the sync service subscription, as we're done consuming everything we need. Will comment and change the var name so it's not rebinding (less confusing).